lib/capitano/resin-toolbox.coffee

```diff
@@ -159,23 +159,30 @@ module.exports =
 
 			console.log(chalk.yellow.bold('* Building..'))
 
-			console.log "- Stopping and Removing any previous '#{appName}' container"
-			docker.stopContainer(appName)
-			.then ->
-				docker.removeContainer(appName)
-			.then ->
-				# Get existing image id and remove it after building new one to preserve build cache
+			console.log "- Stopping and removing any previous '#{appName}' container"
+
+			Promise.all [
+				docker.stopContainer(appName)
+				.then ->
+					docker.removeContainer(appName)
+			,
+				# Get existing image id, to remove it after building new one to preserve build cache
 				docker.inspectImage(appName)
 				.catch (err) ->
 					statusCode = '' + err.statusCode
 					return null if statusCode is '404'
 					throw err
-			.then (oldImageInfo) ->
-				console.log "- Building new '#{appName}' image"
+			,
+				# Get all image ids, to add them as extra possible cache sources
+				docker.getAllImages()
+				.map((image) -> image.Id)
+			]
+			.spread (__, oldImageInfo, existingImageIds) ->
 				docker.buildImage
 					baseDir: baseDir
 					name: appName
 					outStream: process.stdout
+					cacheFrom: existingImageIds
```

Comment:
I'm pretty sure if you don't specify any `cacheFrom`, docker will default to any image in the image store. Means you could drop the call to find out this as well (the image count/gathering can actually take quite a while, on the ARM builders it was talking > 5s for 3000+ images. Embedded hw ofc won't have as many but it's also severely less power).

Reply:
> if you don't specify any cacheFrom, docker will default to any image in the image store.

No, it definitely won't. If I don't pass any `cacheFrom` option to Docker I don't get any caching at all. The [original Moby proposal for `cache-from`](https://github.com/moby/moby/issues/26065) has some explanation of why it can't use all available images by default 

Performance is an interesting point. On a pi3 with just the supervisor + app image it's an unnoticeable amount of time. Is that a representative case? Are we expecting other devices to be much slower, or to have loads more images for some reason?

Reply:
Hmm I was sure it was the other way around, that link is pretty convincing though... Sorry for the noise! Either way given that this is absolutely necessary I think we can disregard the performance aspect.